hips/hip-0020.md

```diff
@@ -0,0 +1,132 @@
+a different result when installed with Helm v4.
+
+In addition to the affects of the changes, Helm v4 development has a fixed timeline and making
+changes to Helm in addition to reworking charts is not likely to fit within that fixed window. Enabling
+the development of charts v3 to happen as an experiment that becomes generally available after
+the release of Helm v4 provides more time to continue the work and get feedback.
+
+The goal is to provide adequate time to work on chart changes while doing it in a way that enables
+trust in existing charts to run as they were tested.
+
+## Rationale
+
+Charts v2 were created for Helm v3 and introduced minor changes. The code that handles the chart
+versions is the same with some checking to handle the differences. This handling ended up having
+numerous bugs that had to be worked out in patch releases.
+
+The chart changes being proposed for Helm v4 are more significant. Mixing those in alongside the
+current chart version handling will have trouble limiting bugs will enabling the changes and keeping
+existing charts functioning properly.
+
+The design specified here is meant to enable the current charts to work as expected while providing
+space for more radical changes.
+
+## Specification
+
+Describe the syntax and semantics of any new feature.
+
+Helm has numerous packages that do various things, from working with charts to storing release
+information. These packages all expect there to be one version of a particular thing. To support
+multiple versions of charts, the `chart`, `chartutil`, and `engine` packages will be made
+multi-version. (e.g., engine will have a v1 and v2 versions). In addition to enabling a new version of
+charts, this will enable the gotpl engine to evolve as needed to support the new version of charts. 
+The `chartutil` may be combined with the `chart` package for simplification of the package structure.
+
+The versioning of the packages will follow the same structure that Kubernetes does with its APIs.
+This means a thing will have a directory and within it will be sub-directories for the versions. The
+version specific implementation will be in the version specific sub-directory. For example,
+
+```
+chart
+├── v2
+└── v3
+```
+
+The version will not use a separate Git repository or Go modules. The reason for this is the added
+complexity of managing the repositories and modules is added work to manage changes which will
+slow down velocity, will make releases more complex, and make the Helm SDK more complicated
+to work with.
+
+The existing versions of these packages will be externally facing while the new versions will be
+developed in `internal` as experiments until they are complete enough for release. When ready for
+release, these packages will be moved to the public locations.
+
+To illustrate this, while in development the `chart` package will have the following structure:
+
+```
+internal/chart
```

Comment:
I suspect a new chart version, will also warrant a new release version.

As such, we will also want `pkg/release/{v2,v3}` 

(Eventually, also code which will turn a chart v3 into a release v3. But I think the structure of that code can be left for another day)

Reply:
Good call. Updated